File: py/3264_FinalArrayStateAfterKMultiplicationOperationsI.py
# 3264. Final Array State After K Multiplication Operations I

from typing import List
import heapq


class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        pq = [[x, i] for i, x in enumerate(nums)]
        heapq.heapify(pq)

        for _ in range(k):
            smallest = heapq.heappop(pq)
            smallest[0] *= multiplier
            nums[smallest[1]] = smallest[0]
            heapq.heappush(pq, smallest)

        return nums


if __name__ == "__main__":
    puzzles = [
        ([2, 1, 3, 5, 6], 5, 2),
        ([1, 2], 3, 4)
    ]
    for puzzle in puzzles:
        print(Solution().getFinalState(*puzzle))

File: go/3264/3264_FinalArrayStateAfterKMultiplicationOperationsI.go
// 3264. Final Array State After K Multiplication Operations I

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"leetcode/utils"
	"os"

	"github.com/emirpasic/gods/trees/binaryheap"
	"github.com/samber/lo"
)

func getFinalState(nums []int, k int, multiplier int) []int {
	pq := binaryheap.NewWith(func(_a, _b interface{}) int {
		a, b := _a.([]int), _b.([]int)
		// if there are multiple occurrences of the minimum value, select the one that appears first
		if a[0] == b[0] {
			return a[1] - b[1]
		}
		// find the minimum value x in nums
		return a[0] - b[0]
	})

	for i, num := range nums {
		pq.Push([]int{num, i})
	}

	for range k {
		_curr, _ := pq.Pop()
		curr := _curr.([]int)
		curr[0] *= multiplier
		nums[curr[1]] = curr[0]
		pq.Push(curr)
	}

	return nums
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 3

	for _, test := range lo.Chunk(lines, testSize) {
		var nums []int
		if err := json.Unmarshal([]byte(test[0]), &nums); err != nil {
			panic(err)
		}
		k := utils.ParseInt(test[1])
		multiplier := utils.ParseInt(test[2])
		result := getFinalState(nums, k, multiplier)
		fmt.Println(result)
	}
}
